### PR TITLE
end-to-end matrix multiplication

### LINKED ISSUE #23: Add matmul example

### PR DESCRIPTION
**Pull request by jorendumoulin**
lets fucking goo 🤘 

This successfully replaces the matmul kernel with a full end-to-end flow, combining the work of a lot of previous PRs

It also uses the actual GEMM hardware instead of the cpu golden model!

I think we can say this resolves #23

### CODE REVIEW

File: compiler/transforms/set_memory_layout.py [hunk 1]
```diff
@@ -44,22 +44,23 @@ def match_and_rewrite(self, linalg_op: linalg.Generic, rewriter: PatternRewriter
                 )
             )
 
-            ## tsl b has an offset of 64 to not collide with the banks of a
+            ## tsl b has an offset of 64 to not collide with the banks of
+            ### a (not yet - need aligned allocation for this)
```
Comment (JosseVanDelm) [inline compiler/transforms/set_memory_layout.py:48]:
```suggestion
            ## TODO: Add offset of 64 to TSL b to not collide with the banks of
            ## a (requires aligned allocation)
```

File: kernels/simple_matmul/main.c [hunk 2]
```diff
@@ -44,103 +44,53 @@ uint32_t strideC = 0;
...
   TwoDMemrefI8_t memrefA;
-  memrefA.data = allocated_a;
+  memrefA.data = &A;
   memrefA.aligned_data = memrefA.data;
-  memrefA.shape[0] = M_size;
-  memrefA.shape[1] = K_size;
-  // The following values of this memref are ignored right now.
-  // A 2D memref is not enough to express a tiled-block layout (=4D),
-  // necessary by the accelerator,
-  // Instead we use the variables strideInnermostA, ldA and strideA
```
Comment (JosseVanDelm) [inline kernels/simple_matmul/main.c:98]: These values still seem to be ignored? Can we leave this comment here in some form maybe?

File: kernels/simple_matmul/main.c [hunk 3]
```diff
@@ -149,7 +99,7 @@ int main() {
 
   int nerr = 0;
   for (int i = 0; i < M_size * N_size; i++) {
-    int32_t error = memrefC.aligned_data[i] - C_golden[i];
```
Comment (JosseVanDelm) [inline kernels/simple_matmul/main.c:152]: Is this change necessary?

File: kernels/simple_matmul/main.c [hunk 4]
```diff
@@ -44,103 +44,53 @@ uint32_t strideC = 0;
...
+  memrefB.data = &B;
+  memrefB.aligned_data = memrefB.data;
   memrefB.offset = 0;
-  memrefB.stride[0] = 0;
-  memrefB.stride[1] = 0;
 
-  TwoDMemrefI32_t memrefC;
-  memrefC.data = allocated_c;
+  TwoDMemrefI8_t memrefC;
+  memrefC.data = &C;
```
Comment (JosseVanDelm) [inline kernels/simple_matmul/main.c:82]: Should this not be i32?

File: compiler/transforms/snax_copy_to_dma.py [hunk 5]
```diff
@@ -276,6 +276,10 @@ def match_and_rewrite(self, op: CopyOp, rewriter: PatternRewriter):
         else:
             dma_loop = remaining_strides.pop(0)
 
+        # if lcb is empty, create init lcb with step 1, bound 1
+        if not lcb:
+            lcb = [Stride(1, 1)]
+
```
Comment (JosseVanDelm) [inline compiler/transforms/snax_copy_to_dma.py:282]: why?

File: compiler/transforms/snax_copy_to_dma.py [hunk 6]
```diff
@@ -192,12 +192,12 @@ def match_and_rewrite(self, op: CopyOp, rewriter: PatternRewriter):
...
             offset = Constant.from_int_and_width(tsl_source.data.offset, IndexType())
             pointer_src = Addi(pointer_src, offset, IndexType())
-            ops_to_insert.append(offset)
+            ops_to_insert.extend([offset, pointer_src])
 
         if tsl_dest.data.offset:
             offset = Constant.from_int_and_width(tsl_dest.data.offset, IndexType())
             pointer_dst = Addi(pointer_dst, offset, IndexType())
-            ops_to_insert.append(offset)
+            ops_to_insert.extend([offset, pointer_dst])
```
Comment (JosseVanDelm) [inline compiler/transforms/snax_copy_to_dma.py:200]: I'm assuming this was previously forgotten?

**Review by JosseVanDelm — COMMENTED** [review]
What a great milestone!

This PR touches on various parts though, and I'm not sure I fully understand where some of the changes come from, and I'd like to make sure I fully understand what's happening :) --> hence all my ugly comments :innocent: 
Can you add some more comments in the PR c):
        else:
        else:
            dma_loop = remaining_strides.pop(0)
            dma_loop = remaining_strides.pop(0)


        # if lcb is empty, create init lcb with step 1, bound 1
        if not lcb:
            lcb = [Stride(1, 1)]

omment itself why certain changes were necessary?

File: compiler/transforms/snax_copy_to_dma.py (hunk 6 above)
Comment (jorendumoulin) [reply compiler/transforms/snax_copy_to_dma.py:200]: image

File: kernels/simple_matmul/main.c (hunk 4 above)
Comment (jorendumoulin) [reply kernels/simple_matmul/main.c:82]: yes, it should!

File: kernels/simple_matmul/main.c (hunk 3 above)
Comment (jorendumoulin) [reply kernels/simple_matmul/main.c:152]: yes! memrefC still has the weird output layout of the accelerator. The DMA transfer then also transforms the data when copying it to C, where it has the same layout as C_golden

File: kernels/simple_matmul/main.c (hunk 2 above)
Comment (jorendumoulin) [reply kernels/simple_matmul/main.c:98]: ok i'll check i maybe removed too much

File: compiler/transforms/set_memory_layout.py (hunk 1 above)
Comment (jorendumoulin) [reply compiler/transforms/set_memory_layout.py:48]: yes, almost done by the way: #96

File: compiler/transforms/snax_copy_to_dma.py (hunk 5 above)
Comment (jorendumoulin) [reply compiler/transforms/snax_copy_to_dma.py:282]: if no common contiguous block is found, the list returns empty. The algorithm however tries to access the list (lcb[-1] a couple lines further). The smallest common contiguous block is always one single element. Maybe better to move this change to the lcb algorithm?

File: compiler/transforms/snax_copy_to_dma.py (hunk 6 above)
Comment (JosseVanDelm) [reply compiler/transforms/snax_copy_to_dma.py:200]: Haha allright

File: compiler/transforms/snax_copy_to_dma.py (hunk 5 above)
Comment (JosseVanDelm) [reply compiler/transforms/snax_copy_to_dma.py:282]: Yes this makes sense! I think it indeed makes more sense to put it there!

File: kernels/simple_matmul/main.c (hunk 3 above)
Comment (JosseVanDelm) [reply kernels/simple_matmul/main.c:152]: And all transforms should have append by the end of leaving function 
`_mlir_ciface_simple_matmul(&memrefA, &memrefB, &memrefC);` no?

File: kernels/simple_matmul/gendata.py [hunk 7]
```diff
@@ -48,8 +48,14 @@ def create_data(file_name: str, variables: dict[str, npt.NDArray]):
...
-    A = np.random.randint(low_bound, high_bound, size=A_size, dtype=np.dtype("int8"))
-    B = np.random.randint(low_bound, high_bound, size=B_size, dtype=np.dtype("int8"))
+    # A = np.random.randint(low_bound, high_bound, size=A_size, dtype=np.dtype("int8"))
+    # B = np.random.randint(low_bound, high_bound, size=B_size, dtype=np.dtype("int8"))
+    A = np.arange(A_size[0] * A_size[1], dtype=np.dtype("int8")).reshape(
+        A_size[0], A_size[1]
+    )
+    B = np.arange(B_size[0] * B_size[1], dtype=np.dtype("int8")).reshape(
+        B_size[0], B_size[1]
+    )
```
Comment (JosseVanDelm) [inline kernels/simple_matmul/gendata.py:58]: maybe change back to randint for the tests

**Review by JosseVanDelm — APPROVED** [review]
If you apply all the suggestions this is perfect to go!
